File: src/main/java/me/nepnep/nepbot/message/command/commands/info/MinecraftServer.java
package me.nepnep.nepbot.message.command.commands.info;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;
import me.nepnep.nepbot.message.command.Category;
import me.nepnep.nepbot.message.command.ICommand;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import okhttp3.*;

import java.io.IOException;
import java.util.List;
import java.util.Map;

public class MinecraftServer implements ICommand {
    @Override
    public void execute(List<String> args, GuildMessageReceivedEvent event) {
        if (args.size() >= 1) {
            TextChannel channel = event.getChannel();
            String address = args.get(0);
            String url = "https://api.mcsrvstat.us/2/" + address;

            OkHttpClient client = new OkHttpClient();
            Request request = new Request.Builder().url(url).build();
            Call call = client.newCall(request);

            try {
                Response response = call.execute();
                ResponseBody body = response.body();

                ObjectMapper mapper = new ObjectMapper();
                EmbedBuilder builder = new EmbedBuilder();
                JSONObject server = mapper.readValue(body.string(), JSONObject.class);

                String cleanMotd = server.getMotd().get("clean").toString().replace(",", "").replace("nig", "no");

                builder.addField("IP/Hostname", String.format("%s:%d/%s", server.getIp(), server.getPort(), server.getHostname()), false);

                builder.addField("MOTD", cleanMotd, false);

                builder.addField("Online players/Max players", String.format("%s/%s", server.getPlayers().get("online"), server.getPlayers().get("max")), false);

                builder.addField("Version", server.getVersion(), false);

                builder.setThumbnail("https://api.mcsrvstat.us/icon/" + address);

                channel.sendMessage(builder.build()).queue();
                response.close();
            } catch (IOException | NullPointerException e) {
                channel.sendMessage("Invalid, offline, or unreachable server").queue();
            }
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class JSONObject {
        String ip;

        String hostname;

        int port;

        Map<String, List<String>> motd;

        Map<String, Object> players;

        String version;

        public String getIp() {
            return ip;
        }

        public String getHostname() {
            return hostname;
        }

        public int getPort() {
            return port;
        }

        public Map<String, List<String>> getMotd() {
            return motd;
        }

        public Map<String, Object> getPlayers() {
            return players;
        }

        public String getVersion() {
            return version;
        }
    }


    @Override
    public String getInvoke() {
        return "mcsrv";
    }

    @Override
    public Permission requiredPermission() {
        return null;
    }

    @Override
    public Category getCategory() {
        return Category.INFO;
    }

    @Override
    public String getDescription() {
        return "Gets a minecraft server's info: ;mcsrv <String address>";
    }
}

File: src/main/kotlin/me/nepnep/nepbot/message/command/commands/info/MinecraftServer.kt
package me.nepnep.nepbot.message.command.commands.info

import dev.minn.jda.ktx.messages.Embed
import me.nepnep.nepbot.mapper
import me.nepnep.nepbot.message.command.AbstractCommand
import me.nepnep.nepbot.message.command.Category
import me.nepnep.nepbot.request
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent

class MinecraftServer : AbstractCommand(
    "mcsrv",
    Category.INFO,
    "Gets a minecraft server's info: ;mcsrv <String address>"
) {
    override suspend fun execute(args: List<String>, event: MessageReceivedEvent, channel: GuildMessageChannel) {
        if (args.size != 1) {
            channel.sendMessage("Invalid usage").queue()
            return
        }

        val server = args[0]
        val url = "https://api.mcsrvstat.us/2/$server"
        event.jda.httpClient.request(url, { response ->
            val json = mapper.readTree(response.body!!.string())

            val ip = json["ip"].textValue()
            val port = json["port"].intValue()
            val hostnameJson = json["hostname"]

            if (hostnameJson == null) {
                channel.sendMessage("Couldn't find server").queue()
                return@request
            }

            val hostname = hostnameJson.textValue()

            val jsonMotd = json["motd"]
            if (jsonMotd == null) {
                channel.sendMessage("Couldn't find server").queue()
                return@request
            }

            val builder = StringBuilder()
            jsonMotd["clean"].elements().forEach {
                builder.append(it.textValue())
            }
            val motd = builder.toString()

            val players = json["players"]
            val online = players["online"].intValue()
            val max = players["max"].intValue()

            val version = json["version"].textValue()

            val embed = Embed { 
                field("IP/Hostname", "$ip:$port/$hostname", false)
                field("MOTD", motd, false)
                field("Online players/Max players", "$online/$max", false)
                field("Version", version, false)
                thumbnail = "https://api.mcsrvstat.us/icon/$server"
            }
            channel.sendMessageEmbeds(embed).queue()
        }) {
            channel.sendMessage("API request failed").queue()
        }
    }
}